homeassistant/components/zha/__init__.py

```diff
@@ -374,6 +381,7 @@ def device_registry(self) -> str:
     def register_entity(self, ieee, entity_obj):
         """Record the creation of a hass entity associated with ieee."""
         self._device_registry[ieee].append(entity_obj)
+        self._entities[entity_obj.entity_id] = entity_obj
```

Comment:
What entities are we storing here? It's against our architecture to store platform entities in a shared container. Those are a platform concern.

Reply:
Ok, what’s the best way to get them then? Relying on looking up the component at runtime by parsing the id is brittle no? If users change entity ids this breaks doesn’t it?

Reply:
I think the problem is that we have conflate the home assistant api (entity) with the zha library api (clusters). The cluster related methods should be removed from the entity to a separate class.

It's ok to store a reference to the entity (entity_id) on the new cluster class and vice versa.

This refactor should be made in a separate PR.

Reply:
If we need to interact with entity (home assistant api) methods from a central place we should use the dispatch helper.

Reply:
funny you mention that... I already did this. it's here if you wanna peek at what I was thinking: https://github.com/dmulcahey/home-assistant/tree/dm/zha-reorganization

Reply:
I’m gonna keep thinking about this... open to any and all suggestions. I’ll close this until I have a better way to do it.

Reply:
Looking some more, I think we should go further. The entity itself shouldn't be the listener. We shouldn't pass the entity to register a listener. Probably use the `ZhaDevice` instance for this instead.

All cluster related things should move to `ZhaDevice`.

Reply:
The entity could listen to some clusters via our dispatch helper signal.